pinot-query-runtime/src/main/java/org/apache/pinot/query/runtime/timeseries/serde/TimeSeriesBlockSerde.java

```diff
@@ -0,0 +1,203 @@
+    TimeBuckets timeBuckets = timeBucketsFromRow(container.get(0));
+    Map<Long, List<TimeSeries>> seriesMap = new HashMap<>();
+    for (int index = 1; index < container.size(); index++) {
+      Object[] row = container.get(index);
+      TimeSeries timeSeries = timeSeriesFromRow(tagNames, row, timeBuckets);
+      long seriesId = Long.parseLong(timeSeries.getId());
+      seriesMap.computeIfAbsent(seriesId, (x) -> new ArrayList<>()).add(timeSeries);
+    }
+    return new TimeSeriesBlock(timeBuckets, seriesMap);
+  }
+
+  public static ByteString serializeTimeSeriesBlock(TimeSeriesBlock timeSeriesBlock)
+      throws IOException {
+    TimeBuckets timeBuckets = Objects.requireNonNull(timeSeriesBlock.getTimeBuckets());
+    List<Object[]> container = new ArrayList<>();
+    DataSchema dataSchema = generateDataSchema(timeSeriesBlock);
+    container.add(timeBucketsToRow(timeBuckets, dataSchema));
+    for (var entry : timeSeriesBlock.getSeriesMap().entrySet()) {
+      for (TimeSeries timeSeries : entry.getValue()) {
+        container.add(timeSeriesToRow(timeSeries, dataSchema));
+      }
+    }
+    TransferableBlock transferableBlock = new TransferableBlock(container, dataSchema, DataBlock.Type.ROW);
+    return DataBlockUtils.toByteString(transferableBlock.getDataBlock());
+  }
+
+  // Internal methods below
+
+  private static DataSchema generateDataSchema(TimeSeriesBlock timeSeriesBlock) {
+    TimeSeries sampledTimeSeries = sampleTimeSeries(timeSeriesBlock).orElse(null);
+    int numTags = sampledTimeSeries == null ? 0 : sampledTimeSeries.getTagNames().size();
+    ColumnDataType[] dataTypes = new ColumnDataType[numTags + 1];
+    String[] columnNames = new String[numTags + 1];
+    for (int tagIndex = 0; tagIndex < numTags; tagIndex++) {
+      columnNames[tagIndex] = sampledTimeSeries.getTagNames().get(tagIndex);
+      dataTypes[tagIndex] = ColumnDataType.STRING;
+    }
+    columnNames[numTags] = "__ts_values";
+    dataTypes[numTags] = ColumnDataType.DOUBLE_ARRAY;
+    return new DataSchema(columnNames, dataTypes);
+  }
+
+  private static List<String> generateTagNames(DataSchema dataSchema) {
+    String[] columnNames = dataSchema.getColumnNames();
+    List<String> tagNames = new ArrayList<>(columnNames.length - 1);
+    for (int index = 0; index < columnNames.length - 1; index++) {
+      tagNames.add(columnNames[index]);
+    }
+    return tagNames;
+  }
+
+  private static Optional<TimeSeries> sampleTimeSeries(TimeSeriesBlock timeSeriesBlock) {
+    if (timeSeriesBlock.getSeriesMap().isEmpty()) {
+      return Optional.empty();
+    }
+    List<TimeSeries> timeSeriesList = timeSeriesBlock.getSeriesMap().values().iterator().next();
+    Preconditions.checkState(!timeSeriesList.isEmpty(), "Found empty time-series list");
+    return Optional.of(timeSeriesList.get(0));
+  }
+
+  private static Object[] timeBucketsToRow(TimeBuckets timeBuckets, DataSchema dataSchema) {
```

Comment:
Might make more sense to move these methods `timeBucketsToRow`, `timeBucketsFromRow`, `timeSeriesToRow` , `timeSeriesFromRow` in an utils class.

Reply:
Right now all of these methods are private hence keeping them here. For the foreseeable future we shouldn't need them outside of this class.